README
phpCrypt 0.4
Ryan Gilfether
http://www.gilfether.com/phpcrypt


ABOUT
=============
phpCrypt is an encryption library written in PHP from scratch. It aims to
implement all major encryption ciphers, modes, and other tools used for
encryption and decryption. phpCrypt does not rely on mCrypt, other PHP
extentions, or PEAR libraries.

It currently support many popular encryption ciphers. In addition it
supports many popular modes of encryption. There are also tools to implement
the different padding schemes, as well as multiple methods of creating an
Initialization Vector (IV) for modes that require one.

phpCrypt is developed by Ryan Gilfether <http://www.gilfether.com/phpcrypt>

phpCrypt is always in development. If you have any questions, found a bug,
or have a feature request, you can contact Ryan at
<http://www.gilfether.com/contact-ryan>


DOCUMENTATION
=============
The phpCrypt website at http://www.gilfether.com/phpcrypt lists much of the
information you need to begin. The phpCrypt website lists all the constants
you need to select ciphers,	modes, padding, and IV methods. In addition,
phpCrypt comes with an `examples` directory which has sample code to help get
you started.

phpCrypt is easy to use. An example of encrypting a string using AES-128
with CTR mode is demonstrated below:

	<?php
	include_once("/path/to/phpcrypt/phpCrypt.php");
	use PHP_Crypt\PHP_Crypt as PHP_Crypt;

	$data = "This is my secret message.";
	$key  = "MySecretKey01234";
	$crypt = new PHP_Crypt($key, PHP_Crypt::CIPHER_AES_128, PHP_Crypt::MODE_CTR);

	$iv = $crypt->createIV();
	$encrypt = $crypt->encrypt($data);

	$crypt->IV($iv);
	$decrypt = $crypt->decrypt($encrypt);
	?>


BYTE PADDING
============
Some modes require a block of data to be padded if it's shorter than than
required block size of the cipher. For example DES encryption works on
8 byte blocks of data. If you a have a block of 6 bytes, then it
may need to be padded 2 bytes before it can be encrypted. Some modes don't
require the data to be padded.

By default phpCrypt uses NULL byte padding when necessary, keeping it
compatible with mCrypt.

phpCrypt has other padding methods available as well. You can specify
the padding you wish to use in two ways. The easiest method is to declare it
in the constructor like so:

	$crypt = new PHP_Crypt($key, PHP_Crypt::CIPHER_AES_128, PHP_Crypt::MODE_CTR, PHP_Crypt::PAD_PKCS7);

Optionally, you can also call the phpCrypt::padding() method to set the padding:

	$crypt = new PHP_Crypt($key, PHP_Crypt::CIPHER_AES_128, PHP_Crypt::MODE_CTR);
	$crypt->padding(PHP_Crypt::PAD_PKCS7);

In the event the padding is set for a mode that does not require padding, the
padding method is ignored. You can get a full list of padding methods available
at http://www.gilfether.com/phpcrypt.

You always have the option of padding the data yourself before sending it
through phpCrypt. In this case you do not need to worry about the phpCrypt
padding methods.

NOTE: NULL byte padding is not stripped off during decryption. This is left for
you to do. phpCrypt can not determine whether a null byte is part of the
original data or was added as padding.


CREATING INITIALIZATION VECTORS
===============================
By default phpCrypt will use the PHP mt_rand() to generate random data used
to create the IV. This method is supported on all operating systems, however
there are more secure ways to generate random data depending on your
operating system

LINUX & UNIX

On Unix based systems, phpCrypt supports reading from /dev/random and
/dev/urandom. This can be done by passing one of the following constants
to phpCrypt::createIV():

	$iv = $crypt->createIV(PHP_Crypt::RAND_DEV_RAND);
	or
	$iv = $crypt->createIV(PHP_Crypt::RAND_DEV_URAND);

MICROSOFT WINDOWS

On Windows systems, you have the option to use the random number generator
found in the Microsoft CAPICOM SDK which is more secure. Before this will
work you must install the Microsoft CAPICOM SDK and enable the PHP `com_dotnet`
extension:

	* Download CAPICOM from Microsoft at
	http://www.microsoft.com/en-us/download/details.aspx?id=25281

	* Double click the MSI file you downloaded and follow the install directions

	* Open a command prompt and register the DLL:
	regsvr32 C:\Program Files\PATH TO\CAPICOM SDK\Lib\X86\capicom.dll

	* Now edit php.ini to enable the com_dotnet extension:
	extension=php_com_dotnet.dll

	* If you are running PHP as an Apache module, restart Apache.

To use the Windows random number generator in CAPICOM you would call createIV()
like so:

	$iv = $crypt->createIV(PHP_Crypt::RAND_WIN_COM);

SUPPLYING YOUR OWN IV

You have the option of creating an IV yourself without using phpCrypt::createIV().
If you wish to create your own IV or use one that was given to you for decryption,
set the IV using phpCrypt::IV() method:

	$crypt->IV($your_custom_iv);

Not all modes require an IV. In the event the IV is set for a mode that does not
require an IV, the IV is ignored. You can get a full list of IV constants and a
list of modes that require an IV at http://www.gilfether.com/phpcrypt


CREATING A KEY
==============
phpCrypt includes a helper function to create a string of random bytes to use as
a key. This can be used in place of setting your own key as shown in the examples
above. The PHP_Crypt::RAND constants are the same as the ones used for the createIV()
function. The second parameter indicates the number of random bytes to create.

	<?php
	$key = PHP_Crypt::createKey(PHP_Crypt::RAND, 16);
	$text = "This is my secret message."
	$crypt = new PHP_Crypt($key, PHP_Crypt::CIPHER_AES_128, PHP_Crypt::MODE_CTR);
	?>

See the examples directory for example code demonstrating the createKey() function.


FULL LIST OF CONSTANTS
======================
For the full list of constants available for Ciphers, Modes, Padding, and IV creation,
visit the phpCrypt website: http://www.gilfether.com/phpcrypt


GPL STUFF
=========
This file is part of phpCrypt

phpCrypt is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

Please read the GPL file included in this distribution for the full license.

CHANGELOG
2014-12-15  phpCrypt-0.5.2  Ryan Gilfether

	* The key is now automatically padded with null bytes if
	it is shorter than required by the cipher.

	* Removed README. Since there was already a README.md,
	the README file was redundant and not needed.

2014-05-09  phpCrypt-0.5.1  Ryan Gilfether

	* Fixed error message in Rijndael.php which inaccurately
	reported an incorrect key size in bits instead of bytes

	* Fixed a typo in the blockSize() call in the Rijdael.php
	contructor

2013-09-18  phpCrypt-0.5  Ryan Gilfether

	* Added CAST-128 cipher

	* Added CAST-256 cipher

	* Added support for 8 and 16 byte keys in addition
	to 24 byte keys for Triple DES (3DES)

	* Removed the OneTimePad cipher. This was redundant since
	the SimpleXOR cipher is the same as a one time pad when
	the key length is the same as the length of data length,
	and the key is unique for each block of data.

	* Added PHP_Crypt::createKey(), a helper function to generate
	a random string of bytes that can be used as a key for a
	cipher. See the examples directory to see how it is used.

	* Added PHP_Crypt::setKey() so that it is possible to set
	the key to a new value without having to create an new
	instance of phpCrypt

	* Renamed PHP_Crypt::cipherSetKey() to PHP_Crypt::cipherKey()

	* Updated the constants used for PHP_Crypt::createKey() and
	PHP_Crypt::createIV(). Both now use the PHP_Crypt::RAND
	constants. The PHP_Crypt::IV constants have been removed.

	* Changed SimpleXOR from a block cipher to a stream cipher
	since this is what it really is.

	* In Cipher.php made the $key member private, renamed
	setKey() to key(), and key() is now used to retrieve the
	current key in use by all extending classes

	* Minor changes to how RC2::splitBytes() works using
	array_map()

	* Fixed an error message that occurs in php 5.3 in RC2.php

	* Renamed Base.php to Core.php

	* Replaced for() loops with array_map() in Core.php where
	possible for minor speed improvements

	* Shortened the name of functions in Core.php, and updated
	all files that use those functions.

	* Create a workaround in Core::rotateBitsLeft32() and
	Core::rotateBitsRight32() which caused them to not always
	work correctly on 32 bit platforms

	* Fixed Core::uInt() and Core::uInt32() which caused them to
	not always work correctly on 32 bit platforms

	* Major bug fix in how CTR mode works. Previously it was not
	fully compatible with mCrypt's CTR mode.

2013-07-21  phpCrypt-0.4  Ryan Gilfether

	* Added the 3-Way cipher. phpCrypt's implementation is
	compatible with mCrypts implementation, though there may
	be a bug in mCrypt's 3-Way gamma() function. phpCrypt
	mimics the bug to make it compatible, if it is a bug.
	Unfortunately I do no know enough about 3-Way nor could
	I find much information on it to know if the gamma() function
	is suppose to behave the way it does. Please read phpCrypt's
	3Way::gamma() function comment for further information.

	* Windows users now have the option to create an IV using
	the secure random number generator found in the Microsoft
	CAPICOM SDK. The new constant is PHP_Crypt::IV_WIN_COM.
	View the README file for more information.

	* Fixed a bug in CTR mode where the last few bytes may
	not get passed to the cipher in some cases

	* Mode::CreateIV() now hashes the random string of bytes
	before returning it as the IV. This helps in the adding
	additional 'randomness' to the IV

	* Bug fixes and minor speed improvements in some of the
	modes.

	* Updated all functions and calculations to handle block
	size in bytes instead of bits. This makes the API less
	confusing and fewer conversions from bits to bytes.

2013-07-19  phpCrypt-0.3.2  Ryan Gilfether

	* Version 0.3.1 contained some incomplete files and bugs
	which were not fixed before the release. This version
	addresses those issues.

	* Fixes in the examples code

	* Removed the PHP_Crypt::setIV() function and renamed to
	PHP_Crypt::IV()

2013-07-18  phpCrypt-0.3.1  Ryan Gilfether

	* Found a major bug in the way IV's are used in Encryption
	and Decryption. The $iv parameter has been removed from
	PHP_Crypt::Encrypt() and PHP_Crypt::Decrypt() function, and
	instead the IV should be set with PHP_Crypt::setIV()

	* Updated the code in the 'examples' directory to reflect
	the changes made to correct the IV bug

	* Updated README and README.md to reflect the changes made
	for the IV bug

2013-07-12  phpCrypt-0.3  Ryan Gilfether

	* Added the Blowfish Cipher

	* Updated CryptBase::decimal2Hex() with a parameter to force
	the returned hex to specified byte size

	* Updated RC2.php to use the uInt() function

	* Added PHP_Crypt::SetIV() function

	* Removed the $iv parameter from PHP_Crypt::CreateIV(),
	now to set the IV use the PHP_Crypt::SetIV() function,
	or pass the IV in the encrypt() or decrypt() function

	* Renamed CryptBase.php to Base.php. Removed it as a parent
	class to all other classes except the Cipher class, as it
	was the only class that really needed it.

	* General performance improvements

2013-06-28  phpCrypt-0.2  Ryan Gilfether

	* Fixed Enigma so that it works on both 64 and 32 bit platforms

	* Renamed & fixed the following functions in CryptBase.php
	to use native PHP function calls:
	uInt(), sInt(), uInt32(), sInt32(), uChar(), sChar()

2013-06-24  phpCrypt-0.1  Ryan Gilfether

	* The initial release of phpCrypt 0.1

Copyright 2013 Ryan Gilfether http://www.gilfether.com/phpcrypt
Copying and distribution of this file, with or without modification,
are permitted provided the copyright notice and this notice are
preserved.

phpCrypt.php
<?php
/*
 * Author: Ryan Gilfether
 * URL: http://www.gilfether.com/phpCrypt
 * Date: Sep 4, 2005
 * Copyright (C) 2005 Ryan Gilfether
 *
 * This file is part of phpCrypt
 *
 * phpCrypt is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

namespace php_crypt;

require_once(dirname(__FILE__)."/Core.php");
require_once(dirname(__FILE__)."/Includes.inc.php");

/**
 * The phpCrypt class, a front end to all the Ciphers and Modes phpCrypt supports
 *
 * @author Ryan Gilfether
 * @link http://www.gilfether.com/phpcrypt
 * @copyright 2005 Ryan Gilfether
 */
class PHP_Crypt
{
	// Ciphers
	const CIPHER_3DES			= "3DES";
	const CIPHER_3WAY			= "3-Way";
	const CIPHER_AES_128		= "AES-128";
	const CIPHER_AES_192		= "AES-192";
	const CIPHER_AES_256		= "AES-256";
	const CIPHER_ARC4			= "ARC4"; // Alternative RC4
	const CIPHER_BLOWFISH		= "Blowfish";
	const CIPHER_CAST_128		= "CAST-128";
	const CIPHER_CAST_256		= "CAST-256";
	const CIPHER_DES			= "DES";
	const CIPHER_ENIGMA			= "Enigma";
	const CIPHER_RC2			= "RC2";
	const CIPHER_RIJNDAEL_128	= "Rijndael-128";
	const CIPHER_RIJNDAEL_192	= "Rijndael-192";
	const CIPHER_RIJNDAEL_256	= "Rijndael-256";
	const CIPHER_SKIPJACK		= "Skipjack";
	const CIPHER_SIMPLEXOR		= "SimpleXOR";
	const CIPHER_VIGENERE		= "Vigenere"; // historical

	// Modes
	const MODE_CBC	= "CBC";
	const MODE_CFB	= "CFB";  // 8 bit cfb mode
	const MODE_CTR	= "CTR";
	const MODE_ECB	= "ECB";
	const MODE_NCFB	= "NCFB"; // blocksize cfb mode
	const MODE_NOFB	= "NOFB"; // blocksize ofb mode
	const MODE_OFB	= "OFB";  // 8 bit ofb mode
	const MODE_PCBC	= "PCBC";
	const MODE_RAW	= "Raw";  // raw encryption, with no mode
	const MODE_STREAM = "Stream"; // used only for stream ciphers

	// The source of random data used to create keys and IV's
	// Used for PHP_Crypt::createKey(), PHP_Crypt::createIV()
	const RAND			= "rand"; // uses mt_rand(), windows & unix
	const RAND_DEV_RAND	= "/dev/random"; // unix only
	const RAND_DEV_URAND= "/dev/urandom";// unix only
	const RAND_WIN_COM	= "wincom";		 // windows only, COM extension
	const RAND_DEFAULT_SZ = 32;			 // the default number of bytes returned

	// Padding types
	const PAD_ZERO			= 0;
	const PAD_ANSI_X923		= 1;
	const PAD_ISO_10126		= 2;
	const PAD_PKCS7			= 3;
	const PAD_ISO_7816_4	= 4;


	/** @type object $cipher An instance of the cipher object selected */
	private $cipher = null;

	/** @type object $mode An instance of the mode object selected */
	private $mode = null;


	/**
	 * Constructor
	 *
	 * @param string $key The key to use for the selected Cipher
	 * @param string $cipher The type of cipher to use
	 * @param string $mode The encrypt mode to use with the cipher
	 * @return void
	 */
	public function __construct($key, $cipher = self::CIPHER_AES_128, $mode = self::MODE_ECB, $padding = self::PAD_ZERO)
	{
		/*
		 * CIPHERS
		 */
		switch($cipher)
		{
		case self::CIPHER_3DES:
			$this->cipher = new Cipher_3DES($key);
			break;

		case self::CIPHER_3WAY:
			$this->cipher = new Cipher_3WAY($key);
			break;

		case self::CIPHER_AES_128:
			$this->cipher = new Cipher_AES_128($key);
			break;

		case self::CIPHER_AES_192:
			$this->cipher = new Cipher_AES_192($key);
			break;

		case self::CIPHER_AES_256:
			$this->cipher = new Cipher_AES_256($key);
			break;

		case self::CIPHER_ARC4: // an alternative to RC4
			$this->cipher = new Cipher_ARC4($key);
			break;

		case self::CIPHER_BLOWFISH:
			$this->cipher = new Cipher_Blowfish($key);
			break;

		case self::CIPHER_CAST_128:
			$this->cipher = new Cipher_CAST_128($key);
			break;

		case self::CIPHER_CAST_256:
			$this->cipher = new Cipher_CAST_256($key);
			break;

		case self::CIPHER_DES:
			$this->cipher = new Cipher_DES($key);
			break;

		case self::CIPHER_ENIGMA:
			$this->cipher = new Cipher_Enigma($key);
			break;

		case self::CIPHER_RC2:
			$this->cipher = new Cipher_RC2($key);
			break;

		case self::CIPHER_RIJNDAEL_128:
			$this->cipher = new Cipher_Rijndael_128($key);
			break;

		case self::CIPHER_RIJNDAEL_192:
			$this->cipher = new Cipher_Rijndael_192($key);
			break;

		case self::CIPHER_RIJNDAEL_256:
			$this->cipher = new Cipher_Rijndael_256($key);
			break;

		case self::CIPHER_SIMPLEXOR:
			$this->cipher = new Cipher_Simple_XOR($key);
			break;

		case self::CIPHER_SKIPJACK:
			$this->cipher = new Cipher_Skipjack($key);
			break;

		case self::CIPHER_VIGENERE:
			$this->cipher = new Cipher_Vigenere($key);
			break;

		default:
			trigger_error("$cipher is not a valid cipher", E_USER_WARNING);
		}


		/*
		 * MODES
		 */
		switch($mode)
		{
		case self::MODE_CBC:
			$this->mode = new Mode_CBC($this->cipher);
			break;

		case self::MODE_CFB:
			$this->mode = new Mode_CFB($this->cipher);
			break;

		case self::MODE_CTR:
			$this->mode = new Mode_CTR($this->cipher);
			break;

		case self::MODE_ECB:
			$this->mode = new Mode_ECB($this->cipher);
			break;

		case self::MODE_NCFB:
			$this->mode = new Mode_NCFB($this->cipher);
			break;

		case self::MODE_NOFB:
			$this->mode = new Mode_NOFB($this->cipher);
			break;

		case self::MODE_OFB:
			$this->mode = new Mode_OFB($this->cipher);
			break;

		case self::MODE_PCBC:
			$this->mode = new Mode_PCBC($this->cipher);
			break;

		case self::MODE_RAW:
			$this->mode = new Mode_RAW($this->cipher);
			break;

		case self::MODE_STREAM:
			$this->mode = new Mode_Stream($this->cipher);
			break;

		default:
			trigger_error("$mode is not a valid mode", E_USER_WARNING);
		}

		// set the default padding
		$this->padding($padding);
	}


	/**
	 * Destructor
	 *
	 * @return void
	 */
	public function __destruct()
	{

	}


	/**
	 * Encrypt a plain text message using the Mode and Cipher selected.
	 * Some stream modes require this function to be called in a loop
	 * which requires the use of $result parameter to retrieve
	 * the decrypted data.
	 *
	 * @param string $text The plain text string
	 * @return string The encrypted string
	 */
	public function encrypt($text)
	{
		// check that an iv is set, if required by the mode
		$this->mode->checkIV();

		// the encryption is done inside the mode
		$this->mode->encrypt($text);
		return $text;
	}


	/**
	 * Decrypt an encrypted message using the Mode and Cipher selected.
	 * Some stream modes require this function to be called in a loop
	 * which requires the use of $result parameter to retrieve
	 * the decrypted data.
	 *
	 * @param string $text The encrypted string
	 * @return string The decrypted string
	 */
	public function decrypt($text)
	{
		// check that an iv is set, if required by the mode
		$this->mode->checkIV();

		// the decryption is done inside the mode
		$this->mode->decrypt($text);
		return $text;
	}


	/**
	 * Return the cipher object being used
	 *
	 * @return object The Cipher object
	 */
	public function cipher()
	{
		return $this->cipher;
	}


	/**
	 * Return the mode object being used
	 *
	 * @return object The Mode object
	 */
	public function mode()
	{
		return $this->mode;
	}


	/**
	 * Returns the name of the cipher being used
	 *
	 * @return string The name of the cipher currently in use,
	 *	it will be one of the predefined phpCrypt cipher constants
	 */
	public function cipherName()
	{
		return $this->cipher->name();
	}


	/**
	 * Return the name of the mode being used
	 *
	 * @return string The name of the mode in use, it will
	 * be one of the predefined phpCrypt mode constants
	 */
	public function modeName()
	{
		return $this->mode->name();
	}


	/**
	 * Returns Ciphers required block size in bytes
	 *
	 * @return integer The cipher data block size, in bytes
	 */
	public function cipherBlockSize()
	{
		return $this->cipher->blockSize();
	}


	/**
	 * Returns the cipher's required key size, in bytes
	 *
	 * @return integer The cipher's key size requirement, in bytes
	 */
	public function cipherKeySize()
	{
		return $this->cipher->keySize();
	}


	/**
	 * Sets and/or returns the key to be used. Normally you set
	 * the key in the phpCrypt constructor. This can be usefully
	 * if you need to change the key on the fly and don't want
	 * to create a new instance of phpCrypt.
	 *
	 * If the $key parameter is not given, this function will simply
	 * return the key currently in use.
	 *
	 * @param string $key Optional, The key to set
	 * @return string The key being used
	 */
	public function cipherKey($key = "")
	{
		return $this->cipher->key($key);
	}


	/**
	 * A helper function which will create a random key. Calls
	 * Core::randBytes(). By default it will use PHP_Crypt::RAND for
	 * the random source of bytes, and return a PHP_Crypt::RAND_DEFAULT_SZ
	 * byte string. There are 4 ways to create a random byte string by
	 * setting the $src parameter:
	 * PHP_Crypt::RAND - Default, uses mt_rand()
	 * PHP_Crypt::RAND_DEV_RAND - Unix only, uses /dev/random
	 * PHP_Crypt::RAND_DEV_URAND - Unix only, uses /dev/urandom
	 * PHP_Crypt::RAND_WIN_COM - Windows only, uses Microsoft's CAPICOM SDK
	 *
	 * @param string $src Optional, The source to use to create random bytes
	 * @param integer $len Optional, The number of random bytes to return
	 * @return string A random string of bytes
	 */
	public static function createKey($src = self::RAND, $len = self::RAND_DEFAULT_SZ)
	{
		return Core::randBytes($src, $len);
	}


	/**
	 * Sets the IV to use. Note that you do not need to call
	 * this function if creating an IV using createIV(). This
	 * function is used when an IV has already been created
	 * outside of phpCrypt and needs to be set. Alternatively
	 * you can just pass the $iv parameter to the encrypt()
	 * or decrypt() functions
	 *
	 * When the $iv parameter is not given, the function will
	 * return the current IV being used. See createIV() if you
	 * need to create an IV.
	 *
	 * @param string $iv Optional, The IV to use during Encryption/Decryption
	 * @return void
	 */
	public function IV($iv = "")
	{
		return $this->mode->IV($iv);
	}


	/**
	 * Creates an IV for the the Cipher selected, if one is required.
	 * If you already have an IV to use, this function does not need
	 * to be called, instead set it with setIV(). If you create an
	 * IV with createIV(), you do not need to set it with setIV(),
	 * as it is automatically set in this function
	 *
	 * $src values are:
	 * PHP_Crypt::RAND - Default, uses mt_rand()
	 * PHP_Crypt::RAND_DEV_RAND - Unix only, uses /dev/random
	 * PHP_Crypt::RAND_DEV_URAND - Unix only, uses /dev/urandom
	 * PHP_Crypt::RAND_WIN_COM - Windows only, uses Microsoft's CAPICOM SDK
	 *
	 * @param string $src Optional, how the IV is generated
	 * @return string The IV that was created, and set for the mode
	 */
	public function createIV($src = self::RAND)
	{
		return $this->mode->createIV($src);
	}


	/**
	 * Sets the type of padding to be used within the specified Mode
	 *
	 * @param string $type One of the predefined padding types
	 * @return void
	 */
	public function padding($type = "")
	{
		return $this->mode->padding($type);
	}
}
?>